### PR TITLE
DV implementation for Kernel

### PR DESCRIPTION
**Pull request by allisonport-db**
Support DVs in Kernel

- Read the DeletionVectorDescriptors in AddFiles
- Read DVs into RoaringBitmapArrays using the FileSystemHandler
- Convert to a selection vector based on row_ids for each DataReadResult

### CODE REVIEW

File: kernel/kernel-api/src/main/java/io/delta/kernel/internal/deletionvectors/RoaringBitmapArray.java [hunk 1]
```diff
@@ -0,0 +1,282 @@
...
+
+    public static RoaringBitmapArray create(long... values) {
+        RoaringBitmapArray bitmap = new RoaringBitmapArray();
+        for (long value : values) {
+            bitmap.add(value);
+        }
+        return bitmap;
+    }
+
+
```
Comment (vkorukanti) [inline kernel/kernel-api/src/main/java/io/delta/kernel/internal/deletionvectors/RoaringBitmapArray.java:281]: nit: empty lines

File: kernel/kernel-default/src/main/java/io/delta/kernel/DefaultKernelUtils.java [hunk 2]
```diff
@@ -47,18 +47,32 @@ public static final MessageType pruneSchema(
...
+                        return prunedType(subType, column.getDataType());
+                    } else {
+                        return null;
+                    }
+                })
+                .filter(Objects::nonNull)
+                .collect(Collectors.toList());
+    }
+
+    private static Type prunedType(Type type, DataType deltaType) {
```
Comment (vkorukanti) [inline kernel/kernel-default/src/main/java/io/delta/kernel/DefaultKernelUtils.java:68]: thanks for fixing this.

**Review by vkorukanti — CHANGES_REQUESTED** [review]
lgtm. Few comments.

File: kernel/kernel-api/src/main/java/io/delta/kernel/internal/deletionvectors/DeletionVectorStoredBitmap.java [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+                                    dvDescriptor.getOffset().orElse(0), // offset
+                                    // we pad 4 bytes in the front for the size
+                                    // and 4 bytes at the end for CRC-32 checksum
+                                    dvDescriptor.getSizeInBytes() + 8 // size
+                            )
+                    );
+
+            CloseableIterator<ByteArrayInputStream> streamIter = fileSystemClient.readFiles(
+                    Utils.singletonCloseableIterator(dvToRead));
+            ByteArrayInputStream stream = InternalUtils.getSingularElement(streamIter).orElseThrow(
```
[the remark this answers is not in the document]
Comment (allisonport-db) [reply kernel/kernel-api/src/main/java/io/delta/kernel/internal/deletionvectors/DeletionVectorStoredBitmap.java:76]: Here it should be closed in `InternalUtils.getSingularElement` is that sufficient? In `loadFromStream` wasn't sure how to do it for both `stream` and `dataStream`

File: kernel/kernel-api/src/main/java/io/delta/kernel/Scan.java [hunk 4]
```diff
@@ -110,14 +122,45 @@ static CloseableIterator<DataReadResult> readData(
...
-                ColumnarBatch updatedBatch =
+            @Override
+            public DataReadResult next()
+            {
+                FileDataReadResult fileDataReadResult = data.next();
+
+                int rowIndexOrdinal = fileDataReadResult.getData().getSchema()
+                    .indexOf(StructField.ROW_INDEX_COLUMN_NAME);
+                ColumnVector rowIndexVector =
+                    fileDataReadResult.getData().getColumnVector(rowIndexOrdinal);
```
[the remark this answers is not in the document]
Comment (allisonport-db) [reply kernel/kernel-api/src/main/java/io/delta/kernel/Scan.java]: Had to rearrange a lot of it but updated

File: kernel/kernel-api/src/main/java/io/delta/kernel/internal/deletionvectors/DeletionVectorStoredBitmap.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (vkorukanti) [reply kernel/kernel-api/src/main/java/io/delta/kernel/internal/deletionvectors/DeletionVectorStoredBitmap.java:76]: you can have something like this:

try(ByteArrayInputStream stream = InternalUtils.getSingularElement(streamIter).orElseThrow(...)) {

}

if you have multiple
```
try(resource r1 = initCode; resource r2 = initCode) {

}
```

File: kernel/kernel-default/src/main/java/io/delta/kernel/parquet/ParquetConverters.java [hunk 5]
```diff
@@ -513,6 +513,7 @@ public ColumnVector getDataColumnVector(int batchSize)
         {
             ColumnVector vector = new DefaultBinaryVector(dataType, batchSize, values);
             // re-initialize the working space
+            this.nullability = initNullabilityVector(nullability.length);
```
Comment (vkorukanti) [inline kernel/kernel-default/src/main/java/io/delta/kernel/parquet/ParquetConverters.java:516]: thank you for fixing this

**Review by vkorukanti — APPROVED** [review]
lgtm.

**Pull request closed without merging**